### PR TITLE
200 js sdk

### LINKED ISSUE #200: Migrate SDK
### Description

This repo currently uses the [Node SDK](https://github.com/sphereio/sphere-node-sdk). In a bid to deprecate this SDK, this module should be refactored to use the [JS SDK](https://github.com/commercetools/nodejs)

### Notes
Support for node v4 will be dropped
This migration will add functionality to update `product-state` via CSV
fyi @Oehmi @daern91 @junajan 

### TODO
- [x] Migrate exporter
- [x] Migrate product state
- [x] Migrate template
- [x] Migrate import
  - [x] Migrate sync

### PR DESCRIPTION
**Pull request by wizzy25**
#### Summary

Migrate this package to the JS SDK

#### Description

This PR replaces the old node SDK with the JS SDK. 

#### Features
- New CSV column: `state`
- Enable [`transitionState`](https://docs.commercetools.com/http-api-projects-products.html#transition-state) update action

#### Breaking Change
- Support for `--sphereProtocol` and `--sphereAuthProtocol` flags have been dropped. Users should pass the full hosts with the `--sphereHost` and `--sphereAuthHost` flags
- Support for node v4 is officially dropped

This is a really large PR and it will be great to get your thoughts. I tried not to change the structure of the code too much as that will entail rewriting the whole package, more or less.

There's still some work to be done - like moving to a more robust logger and using the [`@commercetools/get-credentials`](https://www.npmjs.com/package/@commercetools/get-credentials) package. As this PR is already quite large as it is, these should be reserved for another PR or PRs

resolves #200

### CODE REVIEW

**Comment by tdeekens** [comment]
Sorry, I am the wrong person to review this. I started until 1/2 way down until I released that my feedback is worthless and I can't judge anything here.

File: src/coffee/categories.coffee [hunk 1]
```diff
@@ -15,8 +16,11 @@ class Categories
     @id2fqName = {}
     @duplicateNames = []
 
-  getAll: (client) ->
-    client.categories.all().fetch()
+  getAll: (client, projectKey) ->
```
Comment (butenkor) [inline src/coffee/categories.coffee]: Are you sure you will get all items here?

`client.categories.all().fetch()` from old SDK was aggregating page by page all items for you and was returning only if all items matching the query has been fetched. Unfortunately I did not find equivalent convenience function in nodejs SDK. You could use `client.process` here but then you would need to aggregate all paged results in your callback on your own.

Comment (wizzy25) [reply src/coffee/categories.coffee]: @butenkor Thanks. Modified here ff7eb1573ccd88959978a381fc5268af7e5cf9e5

File: src/spec/validator.spec.coffee [hunk 2]
```diff
@@ -232,8 +232,8 @@ describe 'Validator', ->
         done()
       .catch (err) -> done _.prettify(err)
 
-  xdescribe '#valProduct', ->
-    it 'should return no error', (done) ->
+  describe '#valProduct', ->
+    xit 'should return no error', (done) ->
```
Comment (daern91) [inline src/spec/validator.spec.coffee]: This one probably doesnt belong.

File: src/spec/import.spec.coffee [hunk 3]
```diff
@@ -1,34 +1,31 @@
...
-      expect(importer.client).toBeDefined()
-      expect(importer.client._task._maxParallel).toBe 10
-      expect(importer.sync).toBeDefined()
+      expect(@importer).toBeDefined()
+      expect(@importer.client).toBeDefined()
 
-  xdescribe 'match on custom attribute', ->
-    it 'should find match based on custom attribute', ->
+  describe 'match on custom attribute', ->
+    xit 'should find match based on custom attribute', ->
```
Comment (daern91) [inline src/spec/import.spec.coffee]: Do we wanna have `pending` tests? I see it used to be inside `xdescribe`

Comment (daern91) [reply src/spec/import.spec.coffee]: Got it. What is the reason for using `pending` instead of running them, though?

Comment (wizzy25) [reply src/spec/import.spec.coffee]: Some of the tests don't pass for various reasons (e.g timeout). This PR is not the place to fix these tests as it's already too large. Fixing the tests should be done in another PR :-)

**Comment by wizzy25** [comment]
@butenkor What do you think? Is it good enough now?

**Pull request merged**